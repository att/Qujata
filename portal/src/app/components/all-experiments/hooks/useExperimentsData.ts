import { FetchDataStatus, IHttp, useFetch } from '../../../shared/hooks/useFetch';
import { useEffect, useState } from 'react';
import { APIS } from '../../../apis';
import { useFetchSpinner } from '../../../shared/hooks/useFetchSpinner';
import { useErrorMessage } from '../../../hooks/useErrorMessage';
import { ITestRunResult, ITestRunResultData } from '../../../shared/models/test-run-result.interface';

export type TestRunSubset = Pick<ITestRunResultData, 'id' | 'algorithm' | 'iterations' | 'message_size'>;
export type Experiment = Pick<ITestRunResult, 'id' | 'name' | 'end_time'> & { test_runs: TestRunSubset[] };

export interface IUseExperimentsData {
  test_suites: Experiment[];
  status: FetchDataStatus;
}
export interface ExperimentData {
  id: number;
  name: string;
  algorithms: string[];
  iterations: number[];
  message_sizes: number[];
  end_time: number;
};

export function useExperimentsData(): IUseExperimentsData {
    const [allExperiments, setAllExperiments] = useState<Experiment[]>([]);
    const { get, data, cancelRequest, status, error }: IHttp<Experiment[]> = useFetch({ url: APIS.allExperiments });

    useFetchSpinner(status);
    useErrorMessage(error);
    useEffect(() => {
      get();  
      return cancelRequest;
    }, [get, cancelRequest]);


    useEffect(() => {
      if (data) {
        setAllExperiments(data);
      }
    }, [data, status, allExperiments]);

    return { test_suites: allExperiments, status };
}
