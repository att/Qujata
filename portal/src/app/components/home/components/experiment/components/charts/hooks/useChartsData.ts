import { useEffect, useState } from "react";
import { getKeysOfData, getLabels } from "../utils/chart.utils";
import { ITestRunResultData } from "../../../../../../../shared/models/test-run-result.interface";
import { ILineChartData } from "../models/line-chart-data.interface";
import { colors } from "../../../../../../dashboard/components/charts/LineChart/LineChart.const";
import { IExperimentData } from "../../../Experiment";

export interface IUseChartsData {
    barChartLabels: string[];
    barChartData: ITestRunResultData[];
    barChartKeysOfData: string[];
    lineChartData: ILineChartData;
}

function processedLineChartData(data: ITestRunResultData[], keysOfData: string[]): ILineChartData {
    const processedData = data.reduce((acc: {[key1: string]: {[key2: string]: number[]}}, curr: any) => {
        keysOfData.forEach((key: string) => {
          if (curr.results[key] !== undefined) {
            if (!acc[curr.algorithm]) {
              acc[curr.algorithm] = { [key]: [curr.results[key]] };
            } else {
              if (!acc[curr.algorithm][key]) {
                acc[curr.algorithm][key] = [curr.results[key]];
              } else {
                acc[curr.algorithm][key].push(curr.results[key]);
              }
            }
          }
        });
        return acc;
      }, {});

      const iterations: number[] = data.map((run: ITestRunResultData) => run.iterations);
      iterations.sort((a: number, b: number) => a - b);
      const uniqueIterations: Set<number> = new Set(iterations);

      const chartData = {
        labels:  Array.from(uniqueIterations),
        datasets: Object.keys(processedData).map((algorithm, index) => {
        const data = keysOfData.reduce((acc: {[key: string]: number[]}, key: string) => {
            acc[key] = processedData[algorithm][key];
            return acc;
        }, {});
            return {
                label: `${algorithm}       `,
                data: data,
                fill: false,
                backgroundColor: colors[index % colors.length],
                borderColor: colors[index % colors.length],
                borderWidth: 1,
            };
        })
      };
      
      return chartData;
}

export function useChartsData(props: IExperimentData): IUseChartsData {
    const [barChartLabels, setBarChartLabels] = useState<string[]>([]);
    const [barChartData, setBarChartData] = useState<ITestRunResultData[]>([]);
    const [barChartKeysOfData, setBarChartKeysOfData] = useState<string[]>([]);
    const [lineChartData, setLineChartData] = useState<ILineChartData>();
    
    useEffect(() => {
        if(props.data && props.data.test_runs.length > 0) {
            const testRuns: ITestRunResultData[] = props.data.test_runs;
            setBarChartData(testRuns);
            const labels: string[] = getLabels(testRuns);
            setBarChartLabels(labels);
            const keysOfData: string[] = getKeysOfData(testRuns[0].results);
            setBarChartKeysOfData(keysOfData);
            const lineChartData: ILineChartData = processedLineChartData(testRuns, keysOfData);
            setLineChartData(lineChartData);
        }
    }, [props.data]);

    return {
        barChartLabels,
        barChartData,
        barChartKeysOfData,
        lineChartData,
    } as IUseChartsData;
}
