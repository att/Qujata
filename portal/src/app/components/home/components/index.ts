export * from './experiment';
export * from './latest-experiments';
