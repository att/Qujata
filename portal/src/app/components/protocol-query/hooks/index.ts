export * from './useDuplicateData';
export * from './useGetAlgorithms';
export * from './useGetIterations';
export * from './useMessageSizeData';
