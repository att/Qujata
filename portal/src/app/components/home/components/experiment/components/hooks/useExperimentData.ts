import { useEffect, useState } from "react";
import { APIS } from "../../../../../../apis";
import { replaceParams } from "../../../../../../shared/utils/replaceParams";
import { useParams } from "react-router-dom";
import { ITestRunResult, ITestRunResultData } from "../../../../../../shared/models/test-run-result.interface";
import { TestRunUrlParams } from "../../../../../../shared/models/url-params.interface";
import { FetchDataStatus, IHttp, useFetch } from "../../../../../../shared/hooks/useFetch";
import { sortDataByAlgorithm } from "../charts/utils/test-run.utils";

export interface IUseExperimentData {
    data: ITestRunResultData[];
}

export function useExperimentData(): IUseExperimentData {
  const [testRunData, setTestRunData] = useState<ITestRunResultData[]>([]);
  const { testSuiteId } = useParams<TestRunUrlParams>();
  const testRunUrl: string = replaceParams(APIS.testRunResults, { testSuiteId });
  const { get, data, cancelRequest, status }: IHttp<ITestRunResult> = useFetch({ url: testRunUrl });
  
  useEffect(() => {
      get();
      return cancelRequest;
  }, [get, cancelRequest]);

  useEffect(() => {
      if (status === FetchDataStatus.Success && data) {
          const sortedData: ITestRunResultData[] = sortDataByAlgorithm(data.testRuns);
          setTestRunData(sortedData);
      }
  }, [status, data]);

  return {
      data: testRunData
  } as IUseExperimentData;
}
