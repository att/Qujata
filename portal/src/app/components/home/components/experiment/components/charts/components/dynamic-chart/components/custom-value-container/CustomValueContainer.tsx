import { ValueContainerProps, components } from "react-select"
import { ChartType } from "../../models/dynamic-chart.interface";
import styles from './CustomValueContainer.module.scss';
import { AttSelectOption } from "../../../../../../../../../../shared/components/att-select";
import { PropsWithChildren, useCallback, useRef } from "react";
import cn from 'classnames';
import { capitalizeFirstLetter, getIconByValue } from "../../utils/dynamic-chart.utils";
import { useOutsideClick } from "../../../../../../../../../../hooks/useOutsideClick";

export const CustomValueContainer: React.FC<ValueContainerProps<AttSelectOption, boolean>> = (props: PropsWithChildren<ValueContainerProps<AttSelectOption, boolean>>) => {
    const placeholder: string = props.selectProps?.placeholder as string;
    const inputValue: string = props.selectProps?.inputValue as string;
    const containerRef = useRef(null);

    useOutsideClick(containerRef, () => {
        if (props.selectProps.menuIsOpen) {
          props.selectProps.onMenuClose();
        }
    });

    const handleClick: () => void = useCallback((): void => {
        if (!props.selectProps.menuIsOpen) {
            props.selectProps.onMenuOpen();
        }
    }, [props.selectProps]);

    return (
        <div onClick={handleClick} ref={containerRef}>
            <components.ValueContainer {...props} >
                <div className={styles.input_wrapper}>
                    {props.hasValue && <img className={styles.icon} src={getIconByValue(inputValue as ChartType)} alt={inputValue} />}
                    <span className={cn({[styles.value]: props.hasValue}, {[styles.placeholder]: !props.hasValue})}>{props.hasValue && props.getValue() ? capitalizeFirstLetter(props.getValue()[0]?.value) : placeholder}</span>
                </div>
            </components.ValueContainer>
        </div>
    )
}
  