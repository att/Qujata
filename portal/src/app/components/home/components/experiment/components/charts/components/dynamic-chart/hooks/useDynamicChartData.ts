import { useEffect, useState } from "react";
import { AttSelectOption } from "../../../../../../../../../shared/components/att-select";
import { ITestRunResult } from "../../../../../../../../../shared/models/test-run-result.interface";

export interface IUseDynamicChartData {
  yAxiosOptions: AttSelectOption[];
}
export function useDynamicChartData(chartData: ITestRunResult): IUseDynamicChartData {
  const [yAxiosOptions, setYAxiosOptions] = useState<AttSelectOption[]>([]);

  useEffect(() => {
    const uniqueKeys = new Set<string>();

    for (const testRun of chartData.test_runs) {
      const results = testRun.results;
      for (const key in results) {
        uniqueKeys.add(key);
      }
    }

    if (uniqueKeys.size > 0) {
      setYAxiosOptions(Array.from(uniqueKeys).map(key => ({ label: convertLabelByCapitalLetter(key), value: key })));
    }
  }, [chartData]);

  return {
    yAxiosOptions,
  };
}

function convertLabelByCapitalLetter(str: string): string {
  return str
  .split('_') // split the string by underscore
  .map(word => word.charAt(0).toUpperCase() + word.slice(1)) // capitalize the first letter of each word
  .join(' '); // join the words back together with a space
}
