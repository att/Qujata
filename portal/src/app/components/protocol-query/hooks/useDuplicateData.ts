import { useEffect } from 'react';
import { AttSelectOption } from '../../../shared/components/att-select';
import { ExperimentData } from '../../all-experiments/hooks';

export type DuplicateData = {
  data: ExperimentData | undefined,
  setDuplicateData: (data: any) => void,
  setExperimentName: (name: string) => void,
  setAlgorithms: (options: AttSelectOption[]) => void,
  setIterationsCount: (options: AttSelectOption[]) => void,
  setMessageSize: (options: AttSelectOption[]) => void
}

export const useDuplicateData = (duplicate: DuplicateData) => {
  useEffect(() => {
    if (duplicate.data) {
      const duplicateData = duplicate.data;
      if (duplicateData.name) {
        duplicate.setExperimentName(duplicateData.name);
      }
      if (duplicateData.algorithms) {
        const algorithmOptions = duplicateData.algorithms.map((algorithm: string) => {
          return { label: algorithm, value: algorithm } as AttSelectOption;
        });
        duplicate.setAlgorithms(algorithmOptions);
      }
      if (duplicateData.iterations) {
        const iterationsOptions = duplicateData.iterations.map((iteration: number) => {
          return { label: iteration.toString(), value: iteration.toString() } as AttSelectOption;
        });
        duplicate.setIterationsCount(iterationsOptions);
      }
      if (duplicateData.message_sizes) {
        const messageSizeOptions = duplicateData.message_sizes.map((messageSize: number) => {
          return { label: messageSize.toString(), value: messageSize.toString() } as AttSelectOption;
        });
        duplicate.setMessageSize(messageSizeOptions);
      }
      duplicate.setDuplicateData(undefined);
    }
  }, [duplicate]);
};
