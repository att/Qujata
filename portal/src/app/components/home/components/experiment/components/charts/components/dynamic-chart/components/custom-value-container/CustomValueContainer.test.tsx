import { fireEvent, render, RenderResult } from '@testing-library/react';
import { CustomValueContainer } from './CustomValueContainer';
import { GroupBase, ValueContainerProps } from 'react-select';
import { AttSelectOption } from '../../../../../../../../../../shared/components/att-select';
import { useOutsideClick } from '../../../../../../../../../../hooks/useOutsideClick';
import { useRef } from 'react';

jest.mock('../../../../../../../../../../hooks/useOutsideClick');

describe('CustomValueContainer', () => {
  let mockProps: ValueContainerProps<AttSelectOption<any>, boolean, GroupBase<AttSelectOption<any>>>;

  beforeEach(() => {
    mockProps = {
      children: undefined,
      isDisabled: false,
      clearValue: jest.fn(),
      cx: jest.fn(),
      getStyles: jest.fn(),
      getClassNames: jest.fn(),
      getValue: jest.fn(),
      hasValue: false,
      isMulti: false,
      isRtl: false,
      options: [],
      selectOption: jest.fn(),
      selectProps: { menuIsOpen: false, onMenuOpen: jest.fn(), onMenuClose: jest.fn() } as any,
      setValue: jest.fn(),
      theme: undefined as any
    };
  });

  test('should render CustomValueContainer', () => {
    const { container, getByTestId }: RenderResult = render(<CustomValueContainer {...mockProps} />);

    const valueContainerWrapperElement: HTMLElement = getByTestId('value_container_wrapper');
    fireEvent.click(valueContainerWrapperElement);

    expect(container).toBeTruthy();
  });

  test('should not open the menu when rendering CustomValueContainer', () => {
    mockProps.selectProps.menuIsOpen = true;
    const { container, getByTestId }: RenderResult = render(<CustomValueContainer {...mockProps} />);

    const valueContainerWrapperElement: HTMLElement = getByTestId('value_container_wrapper');
    fireEvent.click(valueContainerWrapperElement);

    expect(container).toBeTruthy();
  });

  test('should activate useOutsideClick hook when the menu is open', () => {
    mockProps.selectProps.menuIsOpen = true;
    (useOutsideClick as jest.Mock).mockImplementation((ref, callback) => {
      document.addEventListener('mousedown', (event) => {
        if (ref.current && !ref.current.contains(event.target)) {
          callback();
        }
      });
    });

    render(<CustomValueContainer {...mockProps} />);
    fireEvent.mouseDown(document);

    expect(mockProps.selectProps.onMenuClose).toHaveBeenCalled();
  });

  test('should activate useOutsideClick hook when the menu is close', () => {
    (useOutsideClick as jest.Mock).mockImplementation((ref, callback) => {
      document.addEventListener('mousedown', (event) => {
        if (ref.current && !ref.current.contains(event.target)) {
          callback();
        }
      });
    });

    render(<CustomValueContainer {...mockProps} />);
    fireEvent.mouseDown(document);

    expect(mockProps.selectProps.onMenuClose).not.toHaveBeenCalled();
  });
});
