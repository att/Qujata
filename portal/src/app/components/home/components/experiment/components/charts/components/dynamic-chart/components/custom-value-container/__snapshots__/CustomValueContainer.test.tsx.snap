// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomValueContainer should render CustomValueContainer 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="input_wrapper"
    >
      <span
        class="placeholder"
      />
    </div>
  </div>
</div>
`;
